### PR TITLE
Add tests for optional attendee handling

### PR DESCRIPTION
**Pull request by anthonyqzhu**
Add tests for optional attendee functionality

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 1]
```diff
@@ -270,5 +272,114 @@ public void notEnoughRoom() {
...
+  public void ignoreBusyOptionalAttendees() {
+    Collection<Event> events = Arrays.asList(
+    new Event("Event 1", TimeRange.fromStartDuration(TIME_0800AM, DURATION_30_MINUTES),
+        Arrays.asList(PERSON_A)),
+    new Event("Event 2", TimeRange.fromStartDuration(TIME_0900AM, DURATION_30_MINUTES),
+        Arrays.asList(PERSON_B)),
+    new Event("Event 3", TimeRange.WHOLE_DAY, Arrays.asList(PERSON_C)));
+
+    MeetingRequest request =
+    new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
```
Comment (yjennywang) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: nit: I would suggest line breaking at the ( after MeetingRequest

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 2]
```diff
@@ -270,5 +272,114 @@ public void notEnoughRoom() {
...
+
+    request.addOptionalAttendee(PERSON_C);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+    Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
```
Comment (yjennywang) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: nit: I would suggest breaking off after asList( (so every TimeRange has its own line)

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 3]
```diff
@@ -270,5 +272,114 @@ public void notEnoughRoom() {
...
+    Collection<Event> events = Arrays.asList(
+    new Event("Event 1", TimeRange.fromStartDuration(TIME_0800AM, DURATION_30_MINUTES),
+        Arrays.asList(PERSON_A)),
+    new Event("Event 2", TimeRange.fromStartDuration(TIME_0900AM, DURATION_30_MINUTES),
+        Arrays.asList(PERSON_B)),
+    new Event("Event 3", TimeRange.fromStartDuration(TIME_0830AM, DURATION_30_MINUTES),
+        Arrays.asList(PERSON_C)));
+
+    MeetingRequest request =
+    new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
```
Comment (yjennywang) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: would suggest reformatting this as well

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 4]
```diff
@@ -34,10 +34,12 @@
   // Some people that we can use in our tests.
   private static final String PERSON_A = "Person A";
   private static final String PERSON_B = "Person B";
+  private static final String PERSON_C = "Person C";
```
Comment (yjennywang) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:37]: what do you think of labeling those fields with MANDATORY or OPTIONAL depending on their characteristic in the test?

that way it's easier to read the tests and immediately determine who is supposed to be mandatory vs. optional.

For tests, I wouldn't bother defining constants -- the tests are isolated so it doesn't actually matter if there's reuse (or no reuse)

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 5]
```diff
@@ -270,5 +272,114 @@ public void notEnoughRoom() {
...
+    new Event("Event 3", TimeRange.fromStartDuration(TIME_0830AM, DURATION_30_MINUTES),
+        Arrays.asList(PERSON_C)));
+
+    MeetingRequest request =
+    new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
+
+    request.addOptionalAttendee(PERSON_C);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
```
Comment (yjennywang) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: reformat with an item on each line

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 6]
```diff
@@ -270,5 +272,114 @@ public void notEnoughRoom() {
...
+    Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+        TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void optionalAttendeeCausesTooShortMeeting() {
```
Comment (yjennywang) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: here and elsewhere, consider being explicit about the success criteria in test case names (e.g. skipPartiallyAvailableAttendee)

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 7]
```diff
@@ -270,5 +272,114 @@ public void notEnoughRoom() {
...
+        new Event("Event 1", TimeRange.WHOLE_DAY,Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.WHOLE_DAY, Arrays.asList(PERSON_B)));
+
+    MeetingRequest request = new MeetingRequest(Arrays.asList(), DURATION_30_MINUTES);
+
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected = Arrays.asList();
```
Comment (yjennywang) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:382]: what does this test case test?

When two attendees are marked optional, do we consider them both mandatory?

If the spec doesn't say, I wonder if it might be better UX to require at least two mandatory attendees. Or just return the entire range (maybe it's a party or a livestream something, no one is required...)

**Comment by yjennywang** [comment]
Some other cases to consider:

- can an attendee be both mandatory and optional?
- boundary conditions. I noticed that all of the tests are for 30 minutes -- do you need something that plays with different timings, or plays with off by one errors (e.g. TimeRange.START_OF_DAY, TIME_0830AM, false) vs. TimeRange.START_OF_DAY, TIME_0830AM, true))
- what happens if there are multiple optional attendees, some of whom can make it but some of whom can't?

**Review by yjennywang — CHANGES_REQUESTED** [review]

**Comment by anthonyqzhu** [comment]
> Some other cases to consider:
> 
> * can an attendee be both mandatory and optional?
> * boundary conditions. I noticed that all of the tests are for 30 minutes -- do you need something that plays with different timings, or plays with off by one errors (e.g. TimeRange.START_OF_DAY, TIME_0830AM, false) vs. TimeRange.START_OF_DAY, TIME_0830AM, true))
> * what happens if there are multiple optional attendees, some of whom can make it but some of whom can't?

1. I would assume it has to be one or the other, and that the MeetingRequest class should handle that case

2. I think the inclusive tag is only used for the end of the day, so it ends on 1440 instead of 1439. Otherwise, non inclusive ranges are used and I guess the tests cover that. I will include some tests with different timings

3. For now, the program returns all times that some optional attendees can attend, doesn't matter how many optional attendees can make it

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 7 above)
Comment (anthonyqzhu) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:382]: In the spec it says to return any time that optional attendees are available when there are no mandatory attendees. For example if optional attendee A is free 8-9AM and optional attendee B is only free 2-3PM with no mandatory attendees, then 8-9AM and 2-3PM are both returned. 

In this particular test case I'm not sure what it is supposed to test  but the spec said to include a test for this scenario

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 4 above)
Comment (anthonyqzhu) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:37]: I think I will keep the names as is because they are used as mandatory and optional in different tests - In this case the constants were already defined in the file so I decided to just reuse them

Comment (yjennywang) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:37]: sounds good, thank you!

**Comment by yjennywang** [comment]
> > Some other cases to consider:
> > 
> > * can an attendee be both mandatory and optional?
> > * boundary conditions. I noticed that all of the tests are for 30 minutes -- do you need something that plays with different timings, or plays with off by one errors (e.g. TimeRange.START_OF_DAY, TIME_0830AM, false) vs. TimeRange.START_OF_DAY, TIME_0830AM, true))
> > * what happens if there are multiple optional attendees, some of whom can make it but some of whom can't?
> 
> 1. I would assume it has to be one or the other, and that the MeetingRequest class should handle that case
> 2. I think the inclusive tag is only used for the end of the day, so it ends on 1440 instead of 1439. Otherwise, non inclusive ranges are used and I guess the tests cover that. I will include some tests with different timings
> 3. For now, the program returns all times that some optional attendees can attend, doesn't matter how many optional attendees can make it

Can you verify that your assumptions for (1) and (2) are true by looking at the documentation and code for MeetingRequest? If it's a plain old java object, then this may not be true (https://en.wikipedia.org/wiki/Plain_old_Java_object).

If those things aren't explicitly validated or required by the contract for MeetingRequest, please add tests to make sure they will be handled correctly.

For 3, can you add a test? (unless I missed an existing test)

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 7 above)
Comment (yjennywang) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:382]: I think this makes sense given the spec. As mentioned in another comment, let's make sure there's something that handles the test case you mentioned in your comment.

**Review by yjennywang — APPROVED** [review]
approving, with requested changes

**Comment by anthonyqzhu** [comment]
> > > Some other cases to consider:
> > > 
> > > * can an attendee be both mandatory and optional?
> > > * boundary conditions. I noticed that all of the tests are for 30 minutes -- do you need something that plays with different timings, or plays with off by one errors (e.g. TimeRange.START_OF_DAY, TIME_0830AM, false) vs. TimeRange.START_OF_DAY, TIME_0830AM, true))
> > > * what happens if there are multiple optional attendees, some of whom can make it but some of whom can't?
> > 
> > 
> > 
> > 1. I would assume it has to be one or the other, and that the MeetingRequest class should handle that case
> > 2. I think the inclusive tag is only used for the end of the day, so it ends on 1440 instead of 1439. Otherwise, non inclusive ranges are used and I guess the tests cover that. I will include some tests with different timings
> > 3. For now, the program returns all times that some optional attendees can attend, doesn't matter how many optional attendees can make it
> 
> Can you verify that your assumptions for (1) and (2) are true by looking at the documentation and code for MeetingRequest? If it's a plain old java object, then this may not be true (https://en.wikipedia.org/wiki/Plain_old_Java_object).
> 
> If those things aren't explicitly validated or required by the contract for MeetingRequest, please add tests to make sure they will be handled correctly.
> 
> For 3, can you add a test? (unless I missed an existing test)

So for 3, I believe the test includeTwoFreeOptionalAttendees handles it, but for the first two assumptions they are not in the documentation. So, I'm not sure whether mandatory attendees can be optional or not, I don't think the tutorial is supposed to have that depth because the request class would have to handle the case of two people having the same name as well. Should I just leave it as there is no defined way to handle this case?

**Comment by yjennywang** [comment]
> > > > Some other cases to consider:
> > > > 
> > > > * can an attendee be both mandatory and optional?
> > > > * boundary conditions. I noticed that all of the tests are for 30 minutes -- do you need something that plays with different timings, or plays with off by one errors (e.g. TimeRange.START_OF_DAY, TIME_0830AM, false) vs. TimeRange.START_OF_DAY, TIME_0830AM, true))
> > > > * what happens if there are multiple optional attendees, some of whom can make it but some of whom can't?
> > > 
> > > 
> > > 
> > > 1. I would assume it has to be one or the other, and that the MeetingRequest class should handle that case
> > > 2. I think the inclusive tag is only used for the end of the day, so it ends on 1440 instead of 1439. Otherwise, non inclusive ranges are used and I guess the tests cover that. I will include some tests with different timings
> > > 3. For now, the program returns all times that some optional attendees can attend, doesn't matter how many optional attendees can make it
> > 
> > 
> > Can you verify that your assumptions for (1) and (2) are true by looking at the documentation and code for MeetingRequest? If it's a plain old java object, then this may not be true (https://en.wikipedia.org/wiki/Plain_old_Java_object).
> > If those things aren't explicitly validated or required by the contract for MeetingRequest, please add tests to make sure they will be handled correctly.
> > For 3, can you add a test? (unless I missed an existing test)
> 
> So for 3, I believe the test includeTwoFreeOptionalAttendees handles it, but for the first two assumptions they are not in the documentation. So, I'm not sure whether mandatory attendees can be optional or not, I don't think the tutorial is supposed to have that depth because the request class would have to handle the case of two people having the same name as well. Should I just leave it as there is no defined way to handle this case?

Oh, how interesting! that's a good observation about two people with the same name, too; that had not occurred to me. :)

Feel free to leave it as is. Bonus points for documenting the expected preconditions in the method javadoc comment. :)
